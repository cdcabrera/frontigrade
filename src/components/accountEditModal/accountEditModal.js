import React from 'react';
import PropTypes from 'prop-types';
import { Alert, Button, Form, Grid, Icon, Modal, Spinner } from 'patternfly-react';
import { connect, reduxActions, reduxTypes, store } from '../../redux/';
import apiTypes from '../../constants/apiConstants';
import helpers from '../../common/helpers';
import { FormField, fieldValidation } from '../formField/formField';

// FixMe: API - Hard Coded Values - "AwsAccount"
class AccountEditModal extends React.Component {
  state = {
    accountName: '',
    accountNameError: null,
    resourceType: 'AwsAccount',
    formTouched: false,
    formValid: false
  };

  // FixMe: API - inconsistent property naming, no underscore in "resourcetype"
  // FixMe: API - inconsistent enum for "resourcetype", account summary response uses "type":"aws" vs "AwsAccount"
  // FixMe: API - inconsistent property naming for ARN, /api/v1/report/accounts/ labels it "arn" vs /api/v1/account/:id/ "account_arn"
  // FixMe: API - patch requires additional property of "resourcetype", an id is also being used...
  // FixMe: API - patch not allowed by preflight, temporarily using put instead
  static getDerivedStateFromProps(props, state) {
    let initialState = null;

    if (!state.formTouched && props.account && props.account[apiTypes.API_RESPONSE_ACCOUNTS_NAME]) {
      initialState = {
        accountName: props.account[apiTypes.API_RESPONSE_ACCOUNTS_NAME],
        accountNameError: ''
      };
    }

    return initialState;
  }

  onCancel = () => {
    this.setState(
      {
        accountName: '',
        accountNameError: null,
        resourceType: 'AwsAccount',
        formTouched: false,
        formValid: false
      },
      () => {
        store.dispatch({
          type: reduxTypes.account.EDIT_ACCOUNT_HIDE
        });
      }
    );
  };

  onChangeAccountName = event => {
    const { value } = event.target;
    const errorMessage = fieldValidation.doesntHaveMinimumCharacters(value, 3)
      ? 'Enter minimum of 3 characters for account name'
      : '';

    this.setState(
      {
        accountName: value,
        accountNameError: errorMessage,
        formTouched: true
      },
      () => this.isFormValid()
    );
  };

  onSubmit = event => {
    const { accountName, resourceType, formValid } = this.state;
    const { account, updateAccount } = this.props;

    event.preventDefault();

    if (formValid) {
      updateAccount(account[apiTypes.API_RESPONSE_ACCOUNTS_ID], {
        [apiTypes.API_SUBMIT_ACCOUNT_NAME]: accountName,
        [apiTypes.API_SUBMIT_ACCOUNT_RESOURCE_TYPE]: resourceType,
        [apiTypes.API_SUBMIT_ACCOUNT_ARN]: account[apiTypes.API_RESPONSE_ACCOUNTS_ARN]
      }).then(
        () => {
          this.onCancel();

          store.dispatch({
            type: reduxTypes.toastNotifications.TOAST_ADD,
            alertType: 'success',
            message: (
              <span>
                Account <strong>{accountName}</strong> successfully updated.
              </span>
            )
          });

          store.dispatch({
            type: reduxTypes.account.UPDATE_ACCOUNTS
          });
        },
        () => {
          this.setState(
            {
              formValid: false
            },
            () => {
              if (!this.props.show) {
                store.dispatch({
                  type: reduxTypes.toastNotifications.TOAST_ADD,
                  alertType: 'error',
                  header: `Error Updating Account`,
                  message: `${this.props.errorMessage}`
                });
              }
            }
          );
        }
      );
    }
  };

  isFormValid() {
    const { accountName, accountNameError } = this.state;
    const { account } = this.props;
    const formValid =
      fieldValidation.isEmpty(accountNameError) && accountName !== account[apiTypes.API_RESPONSE_ACCOUNTS_NAME];

    this.setState({
      formValid
    });
  }

  renderErrorMessage() {
    const { error, errorMessage, pending } = this.props;

    if (error && !pending) {
      return (
        <Alert type="error" onDismiss={this.errorDismissed}>
          <strong>Error</strong> {errorMessage}
        </Alert>
      );
    }

    return null;
  }

  renderPendingMessage() {
    const { pending } = this.props;

    if (pending) {
      return (
        <React.Fragment>
          <Spinner loading size="lg" className="blank-slate-pf-icon" />
          <div className="text-center">Account updating...</div>
        </React.Fragment>
      );
    }

    return null;
  }

  render() {
    const { account, pending, show } = this.props;
    const { accountName, accountNameError, formTouched, formValid } = this.state;
    const displayName = fieldValidation.isEmpty(account[apiTypes.API_RESPONSE_ACCOUNTS_NAME])
      ? account[apiTypes.API_RESPONSE_ACCOUNTS_ID]
      : account[apiTypes.API_RESPONSE_ACCOUNTS_NAME];

    return (
      <Modal show={show} onHide={this.cancel}>
        <Modal.Header>
          <Button className="close" onClick={this.onCancel} aria-hidden="true" aria-label="Close">
            <Icon type="pf" name="close" />
          </Button>
          <Modal.Title>{`Edit Account - ${displayName}`}</Modal.Title>
        </Modal.Header>
        <Modal.Body />
        <Grid fluid>
          {this.renderPendingMessage()}
          {this.renderErrorMessage()}
          {!pending && (
            <Form method="post" horizontal onSubmit={this.onSubmit} autoComplete="off">
              <FormField
                id="edit-account-name"
                label="Account Name"
                error={accountNameError}
                errorMessage={accountNameError}
              >
                <Form.FormControl
                  autoFocus
                  type="text"
                  name="accountName"
                  value={accountName}
                  placeholder="Enter a name for this account"
                  maxLength={256}
                  onChange={this.onChangeAccountName}
                />
              </FormField>
            </Form>
          )}
        </Grid>
        <Modal.Footer>
          <Button bsStyle="default" className="btn-cancel" onClick={this.onCancel}>
            Cancel
          </Button>
          <Button
            bsStyle="primary"
            type="submit"
            onClick={this.onSubmit}
            disabled={pending || !formTouched || !formValid}
          >
            Save
          </Button>
        </Modal.Footer>
      </Modal>
    );
  }
}

AccountEditModal.propTypes = {
  account: PropTypes.object.isRequired,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  pending: PropTypes.bool,
  show: PropTypes.bool.isRequired,
  updateAccount: PropTypes.func
};

AccountEditModal.defaultProps = {
  error: false,
  errorMessage: null,
  pending: false,
  updateAccount: helpers.noop
};

const mapDispatchToProps = dispatch => ({
  updateAccount: (id, data) => dispatch(reduxActions.account.updateAccount(id, data))
});

const mapStateToProps = state => ({ ...state.accountEditModal });

const ConnectedAccountViewEditModal = connect(
  mapStateToProps,
  mapDispatchToProps
)(AccountEditModal);

export { ConnectedAccountViewEditModal as default, ConnectedAccountViewEditModal, AccountEditModal };
