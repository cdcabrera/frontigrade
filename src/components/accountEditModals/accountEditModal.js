import React from 'react';
import PropTypes from 'prop-types';
import { Alert, Button, Form, Grid, Icon, Modal, Spinner } from 'patternfly-react';
import { connect, reduxActions, reduxTypes, store } from '../../redux';
import apiTypes from '../../constants/apiConstants';
import helpers from '../../common/helpers';
import { FormField, fieldValidation } from '../formField/formField';

const initialState = {
  accountId: null,
  accountName: '',
  accountNameError: null,
  displayName: '',
  resourceType: 'AwsAccount',
  formTouched: false,
  formValid: false
};

// FixMe: API - Hard Coded Values - "AwsAccount"
class AccountEditModal extends React.Component {
  state = {
    ...initialState
  };

  onCancel = () => {
    this.setState(
      {
        ...initialState
      },
      () => {
        store.dispatch({
          type: reduxTypes.account.EDIT_ACCOUNT_HIDE
        });
      }
    );
  };

  onChangeAccountName = event => {
    const { value } = event.target;
    const errorMessage = fieldValidation.doesntHaveMinimumCharacters(value, 3)
      ? 'Enter minimum of 3 characters for account name'
      : '';

    this.setState(
      {
        accountName: value,
        accountNameError: errorMessage,
        formTouched: true
      },
      () => this.isFormValid()
    );
  };

  onSubmit = event => {
    const { accountId, accountName, resourceType, formValid } = this.state;
    const { updateAccountField } = this.props;

    event.preventDefault();

    if (formValid) {
      updateAccountField(accountId, {
        [apiTypes.API_SUBMIT_ACCOUNT_NAME]: accountName,
        [apiTypes.API_SUBMIT_ACCOUNT_RESOURCE_TYPE]: resourceType
      }).then(
        () => {
          this.onCancel();

          store.dispatch({
            type: reduxTypes.toastNotifications.TOAST_ADD,
            alertType: 'success',
            message: (
              <span>
                Account <strong>{accountName}</strong> successfully updated.
              </span>
            )
          });

          store.dispatch({
            type: reduxTypes.account.UPDATE_ACCOUNTS
          });
        },
        () => {
          this.setState(
            {
              formValid: false
            },
            () => {
              // eslint-disable-next-line react/destructuring-assignment
              if (Math.floor(this.props.errorStatus / 100) === 5) {
                this.onCancel();
                return;
              }

              // eslint-disable-next-line react/destructuring-assignment
              if (!this.props.show) {
                store.dispatch({
                  type: reduxTypes.toastNotifications.TOAST_ADD,
                  alertType: 'error',
                  header: `Error Updating Account`,
                  message: `${this.props.errorMessage}` // eslint-disable-line react/destructuring-assignment
                });
              }
            }
          );
        }
      );
    }
  };

  // FixMe: API - inconsistent property naming, no underscore in "resourcetype"
  // FixMe: API - inconsistent enum for "resourcetype", account summary response uses "type":"aws" vs "AwsAccount"
  // FixMe: API - inconsistent property naming for ARN, /api/v1/report/accounts/ labels it "arn" vs /api/v1/account/:id/ "account_arn"
  // FixMe: API - patch requires additional property of "resourcetype", an id is also being used...
  static getDerivedStateFromProps(props, state) {
    let updateInitialState = null;

    if (!state.formTouched && props.account && props.account[apiTypes.API_RESPONSE_ACCOUNTS_NAME]) {
      const accountId = props.account[apiTypes.API_RESPONSE_ACCOUNTS_ID];
      const accountName = props.account[apiTypes.API_RESPONSE_ACCOUNTS_NAME];
      const displayName = fieldValidation.isEmpty(accountName) ? accountId : accountName;

      updateInitialState = {
        accountId,
        accountName,
        accountNameError: '',
        displayName
      };
    }

    return updateInitialState;
  }

  isFormValid() {
    const { accountName, accountNameError } = this.state;
    const { account } = this.props;
    const formValid =
      fieldValidation.isEmpty(accountNameError) && accountName !== account[apiTypes.API_RESPONSE_ACCOUNTS_NAME];

    this.setState({
      formValid
    });
  }

  renderErrorMessage() {
    const { error, errorMessage, pending } = this.props;

    if (error && !pending) {
      return (
        <Alert type="error" onDismiss={this.errorDismissed}>
          <strong>Error</strong>: {errorMessage}
        </Alert>
      );
    }

    return null;
  }

  renderPendingMessage() {
    const { pending } = this.props;

    if (pending) {
      return (
        <React.Fragment>
          <Spinner loading size="lg" className="blank-slate-pf-icon" />
          <div className="text-center">Account updating...</div>
        </React.Fragment>
      );
    }

    return null;
  }

  render() {
    const { accountName, accountNameError, displayName, formTouched, formValid } = this.state;
    const { pending, show } = this.props;

    return (
      <Modal show={show} onHide={this.cancel}>
        <Modal.Header>
          <Button className="close" onClick={this.onCancel} aria-hidden="true" aria-label="Close">
            <Icon type="pf" name="close" />
          </Button>
          <Modal.Title>{`Edit Account - ${displayName}`}</Modal.Title>
        </Modal.Header>
        <Modal.Body />
        <Grid fluid>
          {this.renderPendingMessage()}
          {this.renderErrorMessage()}
          {!pending && (
            <Form method="post" horizontal onSubmit={this.onSubmit} autoComplete="off">
              <FormField
                id="edit-account-name"
                label="Account Name"
                error={accountNameError}
                errorMessage={accountNameError}
              >
                <Form.FormControl
                  autoFocus
                  type="text"
                  name="accountName"
                  value={accountName}
                  placeholder="Enter a name for this account"
                  maxLength={256}
                  onChange={this.onChangeAccountName}
                />
              </FormField>
            </Form>
          )}
        </Grid>
        <Modal.Footer>
          <Button bsStyle="default" className="btn-cancel" onClick={this.onCancel}>
            Cancel
          </Button>
          <Button
            bsStyle="primary"
            type="submit"
            onClick={this.onSubmit}
            disabled={pending || !formTouched || !formValid}
          >
            Save
          </Button>
        </Modal.Footer>
      </Modal>
    );
  }
}

AccountEditModal.propTypes = {
  account: PropTypes.object.isRequired,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  errorStatus: PropTypes.number,
  pending: PropTypes.bool,
  show: PropTypes.bool.isRequired,
  updateAccountField: PropTypes.func
};

AccountEditModal.defaultProps = {
  error: false,
  errorMessage: null,
  errorStatus: null,
  pending: false,
  updateAccountField: helpers.noop
};

const mapDispatchToProps = dispatch => ({
  updateAccountField: (id, data) => dispatch(reduxActions.account.updateAccountField(id, data))
});

const mapStateToProps = state => ({ ...state.accountEdit.modal });

const ConnectedAccountEditModal = connect(
  mapStateToProps,
  mapDispatchToProps
)(AccountEditModal);

export { ConnectedAccountEditModal as default, ConnectedAccountEditModal, AccountEditModal };
