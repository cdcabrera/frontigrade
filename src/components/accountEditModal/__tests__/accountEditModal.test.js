import React from 'react';
import configureMockStore from 'redux-mock-store';
import { mount, shallow } from 'enzyme';
import { ConnectedAccountViewEditModal, AccountEditModal } from '../accountEditModal';

describe('AccountEditModal Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a connected component with default props', () => {
    const store = generateEmptyStore({ accountEditModal: { show: true, account: {} } });
    const component = shallow(<ConnectedAccountViewEditModal />, { context: { store } });

    expect(component).toMatchSnapshot('connected');
  });

  it('should render a non-connected component', () => {
    const props = {
      show: true,
      account: {}
    };

    const component = mount(<AccountEditModal {...props} />);
    expect(component.render()).toMatchSnapshot();
  });

  it('should have specific display states defined', () => {
    const props = {
      show: true,
      account: {}
    };

    const component = mount(<AccountEditModal {...props} />);

    component.setProps({
      pending: true
    });
    expect(component.render()).toMatchSnapshot('display pending state');

    component.setProps({
      pending: false,
      error: true
    });
    expect(component.render()).toMatchSnapshot('display error state');
  });

  it('should have specific events defined', () => {
    const props = {
      show: true,
      account: {}
    };
    const component = mount(<AccountEditModal {...props} />);
    const componentInstance = component.instance();

    expect(componentInstance.onCancel).toBeDefined();
    expect(componentInstance.onChangeAccountName).toBeDefined();
    expect(componentInstance.onSubmit).toBeDefined();

    component.find('input[name="accountName"]').simulate('change', { target: { value: '' } });
    expect(componentInstance.state).toMatchSnapshot('expected form error');

    component.find('input[name="accountName"]').simulate('change', { target: { value: 'lorem ipsum' } });
    expect(componentInstance.state).toMatchSnapshot('expected form valid');
  });
});
