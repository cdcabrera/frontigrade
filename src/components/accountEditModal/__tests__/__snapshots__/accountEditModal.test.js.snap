// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountEditModal Component should render a non-connected component 1`] = `
<div
  role="dialog"
>
  <div
    class="fade modal-backdrop in"
  />
  <div
    class="fade in modal"
    role="dialog"
    style="display: block; padding-right: 0px;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
        role="document"
      >
        <div
          class="modal-header"
        >
          <button
            aria-hidden="true"
            aria-label="Close"
            class="close btn btn-default"
            type="button"
          >
            <span
              aria-hidden="true"
              class="pficon pficon-close"
            />
          </button>
          <h4
            class="modal-title"
          >
            Edit Account - undefined
          </h4>
        </div>
        <div
          class="modal-body"
        />
        <div
          class="container-fluid"
        >
          <form
            autocomplete="off"
            class="form-horizontal"
            method="post"
          >
            <div
              class="form-group"
            >
              <label
                class="col-sm-3 control-label"
                for="edit-account-name"
              >
                Account Name
              </label>
              <div
                class="col-sm-9"
              >
                <input
                  class="form-control"
                  id="edit-account-name"
                  maxlength="256"
                  name="accountName"
                  placeholder="Enter a name for this account"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </form>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn-cancel btn btn-default"
            type="button"
          >
            Cancel
          </button>
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Save
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`connected 1`] = `
<AccountEditModal
  account={Object {}}
  error={false}
  errorMessage={null}
  pending={false}
  show={true}
  updateAccountField={[Function]}
/>
`;

exports[`display error state 1`] = `
<div
  role="dialog"
>
  <div
    class="fade modal-backdrop in"
  />
  <div
    class="fade in modal"
    role="dialog"
    style="display: block; padding-right: 0px;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
        role="document"
      >
        <div
          class="modal-header"
        >
          <button
            aria-hidden="true"
            aria-label="Close"
            class="close btn btn-default"
            type="button"
          >
            <span
              aria-hidden="true"
              class="pficon pficon-close"
            />
          </button>
          <h4
            class="modal-title"
          >
            Edit Account - undefined
          </h4>
        </div>
        <div
          class="modal-body"
        />
        <div
          class="container-fluid"
        >
          <div
            class="alert alert-danger"
          >
            <span
              aria-hidden="true"
              class="pficon pficon-error-circle-o"
            />
            <strong>
              Error
            </strong>
             
          </div>
          <form
            autocomplete="off"
            class="form-horizontal"
            method="post"
          >
            <div
              class="form-group"
            >
              <label
                class="col-sm-3 control-label"
                for="edit-account-name"
              >
                Account Name
              </label>
              <div
                class="col-sm-9"
              >
                <input
                  class="form-control"
                  id="edit-account-name"
                  maxlength="256"
                  name="accountName"
                  placeholder="Enter a name for this account"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </form>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn-cancel btn btn-default"
            type="button"
          >
            Cancel
          </button>
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Save
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`display pending state 1`] = `
<div
  role="dialog"
>
  <div
    class="fade modal-backdrop in"
  />
  <div
    class="fade in modal"
    role="dialog"
    style="display: block; padding-right: 0px;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
        role="document"
      >
        <div
          class="modal-header"
        >
          <button
            aria-hidden="true"
            aria-label="Close"
            class="close btn btn-default"
            type="button"
          >
            <span
              aria-hidden="true"
              class="pficon pficon-close"
            />
          </button>
          <h4
            class="modal-title"
          >
            Edit Account - undefined
          </h4>
        </div>
        <div
          class="modal-body"
        />
        <div
          class="container-fluid"
        >
          <div
            class="blank-slate-pf-icon spinner spinner-lg"
          />
          <div
            class="text-center"
          >
            Account updating...
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn-cancel btn btn-default"
            type="button"
          >
            Cancel
          </button>
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Save
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`expected form error 1`] = `
Object {
  "accountName": "",
  "accountNameError": "Enter minimum of 3 characters for account name",
  "formTouched": true,
  "formValid": false,
  "resourceType": "AwsAccount",
}
`;

exports[`expected form valid 1`] = `
Object {
  "accountName": "lorem ipsum",
  "accountNameError": "",
  "formTouched": true,
  "formValid": true,
  "resourceType": "AwsAccount",
}
`;
